servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -164,29 +187,28 @@ public void onNext(final ServiceDiscovererEvent<ResolvedAddress> event) {
                         final List<Host<ResolvedAddress, C>> oldHostsTyped = (List<Host<ResolvedAddress, C>>) oldHosts;
                         if (event.isAvailable()) {
                             if (oldHostsTyped.isEmpty()) {
-                                return singletonList(new Host<>(addr));
+                                return singletonList(createHost(addr));
                             }
-                            final List<Host<ResolvedAddress, C>> newHosts = new ArrayList<>(oldHostsTyped.size() + 1);
+                            final List<Host<ResolvedAddress, C>> newHosts =
+                                    new ArrayList<>(oldHostsTyped.size() + 1);
                             newHosts.addAll(oldHostsTyped);
-                            newHosts.add(new Host<>(addr));
+                            newHosts.add(createHost(addr));
                             return newHosts;
                         } else if (oldHostsTyped.isEmpty()) {
                             return emptyList();
                         } else {
-                            final List<Host<ResolvedAddress, C>> newHosts = new ArrayList<>(oldHostsTyped.size() - 1);
-                            for (int i = 0; i < oldHostsTyped.size(); ++i) {
-                                final Host<ResolvedAddress, C> host = oldHostsTyped.get(i);
-                                if (host.address.equals(addr)) {
-                                    host.markInactive();
-                                    for (int x = i + 1; x < oldHostsTyped.size(); ++x) {
-                                        newHosts.add(oldHostsTyped.get(x));
+                            if (eagerConnectionShutdown) {
+                                return removeFromHostsList(oldHosts, addr);
+                            } else {
+                                for (int i = 0; i < oldHostsTyped.size(); ++i) {
+                                    final Host<ResolvedAddress, C> host = oldHostsTyped.get(i);
```

Comment:
Strange to see an old style indexed for-loop used for this.

Reply:
The pattern from the above condition was replicated, but whilst it made sense above, here it's actually pointless. I replaced it with an enhanced for loop.